examples/offix/test/offline.test.ts

```diff
@@ -0,0 +1,185 @@
+import "cross-fetch/polyfill";
+import "fake-indexeddb/auto";
+
+import gql from "graphql-tag";
+import {
+  ApolloOfflineClient,
+  createClient,
+  NetworkStatus,
+  NetworkStatusChangeCallback
+} from "offix-client";
+import { TestxServer } from "../../../src";
+
+class ToggleableNetworkStatus implements NetworkStatus {
+  private online = true;
+  private callbacks: NetworkStatusChangeCallback[] = [];
+
+  public onStatusChangeListener(callback: NetworkStatusChangeCallback): void {
+    this.callbacks.push(callback);
+  }
+
+  public async isOffline(): Promise<boolean> {
+    return !this.online;
+  }
+
+  public setOnline(online: boolean): void {
+    this.online = online;
+    for (const callback of this.callbacks) {
+      callback.onStatusChange({ online });
+    }
+  }
+}
+
+// TODO: remove this queries once we expose client queries/mutations
+// TODO: from the TestxServer API
+// TODO: https://github.com/aerogear/graphql-testx/issues/15
+export const FIND_ALL_TASKS = gql`
+  query findAllTasks {
+    findAllTasks {
+      id
+      version
+      title
+      description
+      author
+    }
+  }
+`;
+
+export const ADD_TASK = gql`
+  mutation createTask(
+    $version: Int!
+    $description: String!
+    $title: String!
+    $author: String!
+  ) {
+    createTask(
+      input: {
+        version: $version
+        title: $title
+        description: $description
+        author: $author
+      }
+    ) {
+      id
+      version
+      title
+      description
+      author
+    }
+  }
+`;
+
+export const UPDATE_TASK = gql`
+  mutation updateTask(
+    $id: ID!
+    $version: Int!
+    $title: String!
+    $description: String!
+    $author: String!
+  ) {
+    updateTask(
+      id: $id
+      input: {
+        version: $version
+        title: $title
+        description: $description
+        author: $author
+      }
+    ) {
+      id
+      version
+      title
+      description
+      author
+    }
+  }
+`;
+
+describe("Offline mutations", () => {
```

Comment:
So it's only a suggestion but we found that wrapping things up in `describe` blocks makes things more difficult to follow and refactor when needed. Having as much setup and teardown as possible in individual tests tends to be much easier to follow and maintain, even if there is some code duplication.

The main challenge is the global variables that the individual tests are accessing and mofiying all of the place. Two problematic ones were `client` and `networkStatus` which were defined as globals at the top of the `describe` block and then there were random functions and test cases accessing and reassigning those constantly. It made it really difficult to copy/paste a test from one file to another.

I would suggest we do all setup for a test e.g. creating `networkStatus` and `client` within each individual test case. [Here is an example](https://github.com/aerogear/offix/blob/master/packages/offix-scheduler/test/Offix.test.ts#L50-L75). The simpler the better IMO.

We've ended up with a structure in lots of our tests that looks like this.

```ts
// imports
// any global **static** things you need to define here

beforeAll(() => {
  // do some server setup
})

afterAll(() => {
  // cleanup server
})

it('name that describes the test case in detail', () => {
// complete setup and test case contained in here
})

it('test name that describes the exact case in detail', () => {
// complete setup and test case contained in here
})

...
```

We've found that this approach is very simple and easy to follow, more importantly, the less methods and lifecycle methods you use from the testing framework, the easier it would be to migrate/refactor the tests if ever needed.

Reply:
All good points, I try to refactor the test with this new suggestion. Just one question regarding the setup of the server, in that case, we still have to use a global variable because it's better to initialize it only once in the `beforeAll()`, is it right?

Reply:
Yeah there are definitely cases where it's practically unavoidable and the server setup is one of those. But usually you never need to touch the `server` during the actual tests so it's perfectly fine. If you needed to set up a specific server with specific behaviour only required by one test case, I would try to set that up and tear it down within that particular test case.

Reply:
Hi @darahayes, I've updated the tests according to your suggestions, wdyt?